internal/wire/records_test.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wire
+
+import (
+	"bufio"
+	"errors"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+)
+
+// loadRecords gets recursively all .bin files from the recordsPath directory,
+// parses their content to wire Msgs and returns them as an array of testCase
+// structs with headerB and bodyB fields set.
+// If no records are found, it returns an empty slice and no error.
+func loadRecords(recordsPath string) ([]testCase, error) {
+	// Load recursively every file path with ".bin" extension from recordsPath directory
+	var recordFiles []string
+
+	err := filepath.WalkDir(recordsPath, func(path string, entry fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if filepath.Ext(entry.Name()) == ".bin" {
+			recordFiles = append(recordFiles, path)
+		}
+
+		return nil
+	})
+
+	switch {
+	case os.IsNotExist(err):
+		return []testCase{}, nil
```

Comment:
An empty non-nil slice suggests there is something complicated going on, that we need to differentiate between it and the nil slice, but that's not the case there. Let's just return nil.

Reply:
We return nil when there's an error. When file does not exist we don't return an error, so I wanted to differentiate from that by using empty slice. 
It's also more adapted to the usage. If we run the function and check if err != nil, we expect from the result to be useful:

```go
arr, err := fun()
if err != nil { return err }

// here we assume that arr will work with operators, `range`'s, etc.
for _, n := range arr { ... }
```

If we return nil without any error we would need to make another `if` to check if value is correct and (correct me if I'm wrong) it doesn't look like a good practice:

```go
arr, err := fun()
if err != nil { return err }

// if we don't validate the arr it will panic in the next statement
if arr != nil {...}

// here we assume that arr will work with operators, `range`'s, etc.
for _, n := range arr { ... }
```

Reply:
>  suggests there is something complicated going on, that we need to differentiate between it and the nil slice

I understand the point, so I mentioned about this behavior in the comment above the function.

Please let me know what's the final decision :)

Reply:
> If we return nil without any error we would need to make another if to check if value is correct 

That's incorrect; it is perfectly fine to iterate over nil slice: https://go.dev/play/p/CWFD4PXSsfq